rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -704,89 +1156,318 @@ impl MiniBlockScheduler {
-            ranges: self.ranges.clone(),
-            vals_targeted: self.vals_targeted,
+// TODO: Add test cases for the all-preamble and all-trailer cases
+
+// When we schedule a chunk we use the repetition index (or, if none exists, just the # of items
+// in each chunk) to map a user requested range into a set of ChunkInstruction objects which tell
+// us how exactly to read from the chunk.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct ChunkInstructions {
+    // The index of the chunk to read
+    chunk_idx: usize,
+    // A "preamble" is when a chunk begins with a continuation of the previous chunk's list.  If there
+    // is no repetition index there is never a preamble.
+    //
+    // It's possible for a chunk to be entirely premable.  For example, if there is a really large list
+    // that spans several chunks.
+    preamble: PreambleAction,
+    // How many complete rows (not including the preamble or trailer) to skip
+    //
+    // If this is non-zero then premable must not be Take
+    rows_to_skip: u64,
+    // How many complete (non-preamble / non-trailer) rows to take
+    rows_to_take: u64,
+    // A "trailer" is when a chunk ends with a partial list.  If there is no repetition index there is
+    // never a trailer.
+    //
+    // It's possible for a chunk to be entirely trailer.  This would mean the chunk starts with the beginning
+    // of a list and that list is continued in the next chunk.
+    //
+    // If this is true then we want to include the trailer
+    take_trailer: bool,
+}
+
+// First, we schedule a bunch of [`ChunkInstructions`] based on the users ranges.  Then we
+// start decoding them, based on a batch size, which might not align with what we scheduled.
+//
+// This results in `ChunkDrainInstructions` which targets a contiguous slice of a `ChunkInstructions`
+//
+// So if `ChunkInstructions` is "skip preamble, skip 10, take 50, take trailer" and we are decoding in
+// batches of size 10 we might have a `ChunkDrainInstructions` that targets that chunk and has its own
+// skip of 17 and take of 10.  This would mean we decode the chunk, skip the preamble and 27 rows, and
+// then take 10 rows.
+//
+// One very confusing bit is that `rows_to_take` includes the trailer.  So if we have:
+// "no preamble, skip 5, take 10, take trailer" and we are draining 20 rows then the
+// first drain instructions will have no preamble, skip 0, take 11 and the second chunk
+// instructions will have take preamble, skip 0, take 9 (assuming the second chunk has at
+// least 9 rows)
+#[derive(Debug, PartialEq, Eq)]
+struct ChunkDrainInstructions {
+    chunk_instructions: ChunkInstructions,
+    rows_to_skip: u64,
+    rows_to_take: u64,
+    preamble_action: PreambleAction,
+}
+
+impl ChunkInstructions {
+    // Given a repetition index and a set of user ranges we need to figure out how to read from the chunks
+    //
+    // We assume that `user_ranges` are in sorted order and non-overlapping
```

Comment:
Should we add a `debug_assert` for this?

Reply:
Maybe but we should probably put it higher up in `decoder.rs`.  This is an assumption we make throughout the decoders and many of them rely on it.

This comment was less "oh, in this spot we have this extra guarantee" and more "remember that we know this"

Reply:
If you have something higher up that is fine.